### PR TITLE
Add shared_directories preset parameter

### PR DESCRIPTION
**Pull request by jamesmcm**
Created new PR for this since I had to solve merge conflicts with preset directories PR #33 

Now using MountStack for automatic unmounting which works much, much better than the previous attempt - thanks for the heads up (+ the nix crate is awesome!).

The shared_directories preset parameter allows some directories to be mounted during installation (i.e. to transfer images or dotfiles, etc.)

See presets/copy_file.toml for an example.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -138,6 +138,7 @@ fn create(command: args::CreateCommand) -> Result<(), Error> {
     let genfstab = Tool::find("genfstab")?;
     let mkfat = Tool::find("mkfs.fat")?;
     let mkext4 = Tool::find("mkfs.ext4")?;
+    let mkdir = Tool::find("mkdir")?;
```
Comment (r-darwish) [inline src/main.rs]: No need external mkdir. Rust has [a function](https://doc.rust-lang.org/std/fs/fn.create_dir_all.html) for that.

File: src/presets.rs [hunk 2]
```diff
@@ -51,19 +55,52 @@ impl Preset {
...
+                        x.iter()
+                            .cloned()
+                            .map(|y| {
+                                let full_path = path.parent().unwrap().join(&y);
+                                if full_path.is_dir() {
+                                    Ok(full_path)
+                                } else {
+                                    Err(ErrorKind::Preset(format!(
+                                        "Preset: {} - shared directory: {} is not directory",
+                                        path.to_string_lossy(),
```
Comment (r-darwish) [inline src/presets.rs]: paths have `.display()`

File: src/storage/mount_stack.rs [hunk 3]
```diff
@@ -38,6 +38,24 @@ impl<'a> MountStack<'a> {
...
+        source: PathBuf,
+        target: PathBuf,
+        options: Option<&str>,
+    ) -> nix::Result<()> {
+        debug!("Mounting {:?} to {:?}", source, target);
+        mount::<_, _, str, _>(
+            Some(&source),
+            &target,
+            None,
+            MsFlags::MS_BIND | MsFlags::MS_NOATIME | MsFlags::MS_RDONLY,
```
Comment (r-darwish) [inline src/storage/mount_stack.rs]: if `MS_RDONLY` indeed doesn't work then we should remove it to avoid confusion

**Review by r-darwish — CHANGES_REQUESTED** [review]
Thank you. There are some small changes to make and we'll be good to go

File: src/storage/mount_stack.rs (hunk 3 above)
Comment (jamesmcm) [reply src/storage/mount_stack.rs]: Apparently it works now: https://unix.stackexchange.com/questions/128336/why-doesnt-mount-respect-the-read-only-option-for-bind-mounts

I'll try to test this directly.

Comment (jamesmcm) [reply src/storage/mount_stack.rs]: Okay, it does not mount as read-only. So I removed the flag.

**Comment by jamesmcm** [comment]
@r-darwish is this okay to merge now? As my own presets are dependent on it 😄

**Comment by r-darwish** [comment]
Thanks

**Pull request merged**